features/370-auth-ui/design.md

```diff
@@ -0,0 +1,393 @@
+# Auth UI
+
+Table of Content
+
+* [Background](#background)
+* [Where Auth UI is served](#where-auth-ui-is-served)
+   * [Reason 1: Individual gears should be unaware of Custom Domain](#reason-1-individual-gears-should-be-unaware-of-custom-domain)
+   * [Reason 2: Cookies written by Auth Gear is host-only](#reason-2-cookies-written-by-auth-gear-is-host-only)
+   * [The actual location of Auth UI](#the-actual-location-of-auth-ui)
+* [API key as client_id](#api-key-as-client_id)
+* [Page templates](#page-templates)
+   * [Customization of page templates](#customization-of-page-templates)
+      * [Customizing the theme with CSS](#customizing-the-theme-with-css)
+      * [Customizing with own templates](#customizing-with-own-templates)
+* [The authorize endpoint](#the-authorize-endpoint)
+   * [The authorization page](#the-authorization-page)
+   * [The signup page](#the-signup-page)
+   * [The signup password page](#the-signup-password-page)
+* [The settings endpoint](#the-settings-endpoint)
+* [The verification endpoint](#the-verification-endpoint)
+* [The logout endpoint](#the-logout-endpoint)
+* [Omitted features](#omitted-features)
+* [Future improvements](#future-improvements)
+   * [Customizing the signup flow or the authentication flow](#customizing-the-signup-flow-or-the-authentication-flow)
+* [Support Matrix](#support-matrix)
+* [Being a OpenID Provider](#being-a-openid-provider)
+   * [Authorization Endpoint](#authorization-endpoint)
+   * [Token Endpoint](#token-endpoint)
+   * [Token Response](#token-response)
+* [Security considerations](#security-considerations)
+   * [Being embedded in an HTML iframe](#being-embedded-in-an-html-iframe)
+* [Appendix](#appendix)
+
+## Background
+
+Auth UI is the builtin web frontend of Auth Gear.
+It provides access to most Auth Gear features as a server side rendered web application.
+It is part of Auth Gear.
+
+It is intended to be a conforming OpenID Provider implementation of Auth Gear in the future.
+As of the time of writing, the conformance of OpenID Provider is not a goal.
+However, the implementation must be prepared for conformance in the future.
+
+## Where Auth UI is served
+
+It is served in the same domain as Auth Gear and Microservices.
+
+### Reason 1: Individual gears should be unaware of Custom Domain
+
+Auth UI is going to be responsible for user verification. The verification email includes an URL to Auth UI. This implies Auth Gear has to know where Auth UI is served. When Custom Domain was introduced, it is a feature solely handled by Gateway. Individual gears such as Auth Gear are unaware of Custom Domain. Serving Auth UI in a different domain will break this property. If Auth UI is served in the same domain, it is trivial for Auth Gear to know where it is.
```

Comment:
> The verification email includes an URL to Auth UI.

I'm thinking maybe it is not a must? Auth Gear itself already have a feature to display the page after verification success, maybe we can just reuse that instead of having "yet another customizable page" on auth UI?